mcfly/modelgen.py

```diff
@@ -115,630 +60,64 @@ def generate_models(x_shape,
+
+    # Replace default hyperparameter ranges with input
+    for key, value in hyperparameter_ranges.items():
+        if key in defaults:
+            print("The value of {} is set from {} (default) to {}".format(key, defaults[key], value))
+            defaults[key] = value
+
+    # Add missing parameters from default
+    for key, value in defaults.items():
+        if key not in hyperparameter_ranges:
+            hyperparameter_ranges[key] = value
+
     model_types_selected = []
-    model_types = settings['model_types']
 
     for i in range(int(np.ceil(number_of_models/len(model_types)))):
         np.random.shuffle(model_types)
         model_types_selected.extend(model_types)
 
-    # Create list of Keras models
+    # Create list of Keras models and their hyperparameters
     # -------------------------------------------------------------------------
     models = []
-    for i in range(0, number_of_models):
+    for i in range(number_of_models):
         current_model_type = model_types_selected[i]
-        generate_model = None
-        if current_model_type == 'CNN':
-            generate_model = generate_CNN_model  # generate_model is a function
-            hyperparameters = generate_CNN_hyperparameter_set(settings)
-        if current_model_type == 'DeepConvLSTM':
-            generate_model = generate_DeepConvLSTM_model
-            hyperparameters = generate_DeepConvLSTM_hyperparameter_set(settings)
-        if current_model_type == 'ResNet':
-            generate_model = generate_resnet_model
-            hyperparameters = generate_resnet_hyperparameter_set(settings)
-        if current_model_type == 'InceptionTime':
-            generate_model = generate_InceptionTime_model
-            hyperparameters = generate_InceptionTime_hyperparameter_set(settings)
-
-        models.append(
-            (generate_model(x_shape, 
-                            number_of_classes, 
-                            metrics=settings['metrics'], **hyperparameters),
-             hyperparameters, current_model_type))
+        if current_model_type in default_models:
+            model_type = default_models[current_model_type](x_shape, number_of_classes,
+                                                            metrics, **hyperparameter_ranges)
+            hyperparameters = model_type.generate_hyperparameters()
+            model = model_type.create_model(**hyperparameters)
+            model_name = current_model_type
+        elif isinstance(current_model_type, str):
+            raise NameError("Unknown model name given.")
+        else: # Assume model class was passed
+            model_type = current_model_type(x_shape, number_of_classes,
+                                            metrics, **hyperparameter_ranges)
+            hyperparameters = model_type.generate_hyperparameters()
+            model = model_type.create_model(**hyperparameters)
+            model_name = model_type.model_name
```

Comment:
This case could be unified with the first case, most of the logic is shared

Reply:
Good point. I removed the redundant lines.